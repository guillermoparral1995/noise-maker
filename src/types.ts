import type { ActionTypes as EnvelopeActionTypes } from './components/Controls/EnvelopeControls/store/actions';
import type { ActionTypes as GeneralControlsActionTypes } from './components/Controls/GeneralControls/store/actions';
import type { ActionTypes as FilterActionTypes } from './components/Controls/FilterControls/store/actions';
import type { ActionTypes as LFOActionTypes } from './components/Controls/LFOControls/store/actions';
import { LFOTargetAction } from './providers/AudioContextProvider';

export type ActionTypes =
  | EnvelopeActionTypes
  | GeneralControlsActionTypes
  | FilterActionTypes
  | LFOTargetAction
  | LFOActionTypes;

export enum Waveform {
  SINE = 'sine',
  SQUARE = 'square',
  SAWTOOTH = 'sawtooth',
  TRIANGLE = 'triangle',
}

export enum FilterType {
  LOWPASS = 'lowpass',
  HIGHPASS = 'highpass',
}

export enum Selectors {
  WAVEFORM = 'waveform',
  FILTER = 'filter',
  LFO_TARGET = 'lfoTarget',
  LFO_WAVEFORM = 'lfoWaveform',
}

export enum Knobs {
  VOLUME = 'VOLUME',
  PAN = 'PAN',
  DETUNE = 'DETUNE',
  ATTACK = 'ATTACK',
  DECAY = 'DECAY',
  SUSTAIN = 'SUSTAIN',
  RELEASE = 'RELEASE',
  FILTER_CUTOFF = 'FILTER_CUTOFF',
  FILTER_RESONANCE = 'FILTER_RESONANCE',
  LFO_FREQUENCY = 'LFO_FREQUENCY',
  LFO_AMPLITUDE = 'LFO_AMPLITUDE',
}

export type LFOTarget =
  | Exclude<
      Knobs,
      | Knobs.LFO_FREQUENCY
      | Knobs.LFO_AMPLITUDE
      | Knobs.ATTACK
      | Knobs.DECAY
      | Knobs.SUSTAIN
      | Knobs.RELEASE
    >
  | 'off';

export interface State {
  volume: number;
  pan: number;
  waveform: Waveform;
  attack: number;
  decay: number;
  sustain: number;
  release: number;
  filter: {
    type: FilterType;
    cutoff: number;
    resonance: number;
  };
  lfo: {
    target: LFOTarget;
    waveform: Waveform;
    frequency: number;
    amplitude: number;
  };
}

export enum Actions {
  UPDATE_VOLUME,
  UPDATE_PAN,
  UPDATE_WAVEFORM,
  UPDATE_DETUNE,
  UPDATE_ATTACK,
  UPDATE_DECAY,
  UPDATE_SUSTAIN,
  UPDATE_RELEASE,
  UPDATE_FILTER_TYPE,
  UPDATE_FILTER_CUTOFF,
  UPDATE_FILTER_RESONANCE,
  UPDATE_LFO_TARGET,
  UPDATE_LFO_WAVEFORM,
  UPDATE_LFO_FREQUENCY,
  UPDATE_LFO_AMPLITUDE,
}
export type ActionBuilder<T> = (payload: T) => ActionTypes;
