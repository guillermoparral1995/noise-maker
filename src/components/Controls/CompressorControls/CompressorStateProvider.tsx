import React, { PropsWithChildren, useReducer } from 'react';
import { ActionTypes } from './store/actions';
import initialState, { CompressorState } from './store/initialState';
import reducer from './store/reducer';

interface CompressorStateProviderValue {
  state: CompressorState;
  dispatch: React.Dispatch<ActionTypes>;
}

export const compressorStateContext =
  React.createContext<CompressorStateProviderValue>(undefined);

export const CompressorStateProvider = ({
  children,
  __mockDispatch,
}: { __mockDispatch?: React.Dispatch<ActionTypes> } & PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <compressorStateContext.Provider
      value={{ state, dispatch: __mockDispatch ?? dispatch }}
    >
      {children}
    </compressorStateContext.Provider>
  );
};
