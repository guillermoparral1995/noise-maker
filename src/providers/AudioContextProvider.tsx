import React, { PropsWithChildren, useEffect, useMemo } from 'react';

interface AudioContextProviderValue {
  context: AudioContext;
  volume: GainNode;
  pan: StereoPannerNode;
  filter: BiquadFilterNode;
  output: AudioNode;
}

export const audioContext =
  React.createContext<AudioContextProviderValue>(undefined);

export const AudioContextProvider = ({ children }: PropsWithChildren) => {
  const context = useMemo(() => new AudioContext(), []);
  const volumeNode: GainNode = useMemo(() => new GainNode(context), []);
  const pannerNode: StereoPannerNode = useMemo(
    () => new StereoPannerNode(context),
    [],
  );
  const filterNode: BiquadFilterNode = useMemo(
    () => new BiquadFilterNode(context),
    [],
  );

  useEffect(() => {
    filterNode.connect(volumeNode);
    volumeNode.connect(pannerNode);
    pannerNode.connect(context.destination);
  }, []);

  return (
    <audioContext.Provider
      value={{
        context,
        volume: volumeNode,
        pan: pannerNode,
        filter: filterNode,
        output: filterNode,
      }}
    >
      {children}
    </audioContext.Provider>
  );
};
