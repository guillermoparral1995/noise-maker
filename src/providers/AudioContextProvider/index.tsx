import React, {
  PropsWithChildren,
  useEffect,
  useMemo,
  useReducer,
} from 'react';
import {
  AnalyserNodeMock,
  AudioContextMock,
  AudioNodeMock,
  BiquadFilterNodeMock,
  GainNodeMock,
  StereoPannerNodeMock,
} from '../../../__mocks__';
import { ActionTypes, LFO1Target, LFO2Target } from '../../types';
import { initialState } from './store/initialState';
import { reducer } from './store/reducer';

export interface LFO<T> {
  output: GainNode;
  target: T;
}

export interface LFOMock<T> {
  output: GainNodeMock;
  target: T;
}

interface AudioContextProviderMocks {
  context: AudioContextMock;
  volume?: GainNodeMock;
  pan?: StereoPannerNodeMock;
  filter?: BiquadFilterNodeMock;
  lfo1?: LFOMock<LFO1Target>;
  lfo2?: LFOMock<LFO2Target>;
  analyser?: AnalyserNodeMock;
  dispatch?: React.Dispatch<ActionTypes>;
  output?: AudioNodeMock;
}

interface AudioContextProviderParams {
  context: AudioContext;
  volume: GainNode;
  pan: StereoPannerNode;
  filter: BiquadFilterNode;
  lfo1: LFO<LFO1Target>;
  lfo2: LFO<LFO2Target>;
  analyser: AnalyserNode;
  dispatch: React.Dispatch<ActionTypes>;
  output: AudioNode;
}

type AudioContextProviderValue = { __isMock: boolean } & (
  | AudioContextProviderParams
  | AudioContextProviderMocks
);

export const audioContext =
  React.createContext<AudioContextProviderValue>(undefined);

export const AudioContextProvider = ({
  children,
  __mocks,
}: { __mocks?: AudioContextProviderMocks } & PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);
  if (__mocks?.context) {
    return (
      <audioContext.Provider
        value={{
          __isMock: true,
          context: __mocks.context,
          volume: __mocks.volume,
          pan: __mocks.pan,
          filter: __mocks.filter,
          lfo1: {
            output: __mocks.lfo1?.output,
            target: __mocks.lfo1?.target,
          },
          lfo2: {
            output: __mocks.lfo2?.output,
            target: __mocks.lfo2?.target,
          },
          analyser: __mocks.analyser,
          output: __mocks.output,
          dispatch: __mocks.dispatch,
        }}
      >
        {children}
      </audioContext.Provider>
    );
  }
  const context = useMemo(() => new AudioContext(), []);
  const volumeNode: GainNode = useMemo(() => new GainNode(context), []);
  const pannerNode: StereoPannerNode = useMemo(
    () => new StereoPannerNode(context),
    [],
  );
  const filterNode: BiquadFilterNode = useMemo(
    () => new BiquadFilterNode(context),
    [],
  );
  const analyser = useMemo(
    () => new AnalyserNode(context, { fftSize: 512 }),
    [],
  );
  const lfo1OutputNode: GainNode = useMemo(() => new GainNode(context), []);
  const lfo2OutputNode: GainNode = useMemo(() => new GainNode(context), []);

  useEffect(() => {
    filterNode.connect(volumeNode);
    volumeNode.connect(pannerNode);
    pannerNode.connect(analyser);
    analyser.connect(context.destination);
  }, []);

  return (
    <audioContext.Provider
      value={{
        __isMock: false,
        context,
        volume: volumeNode,
        pan: pannerNode,
        filter: filterNode,
        lfo1: {
          output: lfo1OutputNode,
          target: state.lfo1Target,
        },
        lfo2: {
          output: lfo2OutputNode,
          target: state.lfo2Target,
        },
        analyser,
        output: filterNode,
        dispatch,
      }}
    >
      {children}
    </audioContext.Provider>
  );
};
