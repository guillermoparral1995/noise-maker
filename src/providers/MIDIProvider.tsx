import React, { PropsWithChildren, useEffect, useState } from 'react';
import { WebMidi } from 'webmidi';

export const midiContext = React.createContext(undefined);

export const MIDIProvider = ({ children }: PropsWithChildren) => {
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState(false);

  const enableMidi = async () => {
    try {
      await WebMidi.enable();
      setLoading(false);
    } catch (e) {
      console.error('Could not start Web MIDI', e);
      setLoading(false);
      setError(true);
    }
  };

  useEffect(() => {
    enableMidi();
  }, []);

  if (loading) return;
  if (!loading && !error)
    return (
      <midiContext.Provider
        value={WebMidi.getInputByName('Arturia MiniLab mkII')}
      >
        {children}
      </midiContext.Provider>
    );
  return children;
};
