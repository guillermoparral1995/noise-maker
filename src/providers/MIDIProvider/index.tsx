import React, { PropsWithChildren, useEffect, useReducer } from 'react';
import { Input, WebMidi } from 'webmidi';
import {
  ActionTypes,
  updateMIDIError,
  updateMIDIInput,
  updateMIDILoading,
} from './store/actions';
import { initialState } from './store/initialState';
import { reducer } from './store/reducer';

interface MIDIContextValue {
  dispatch: React.Dispatch<ActionTypes>;
  inputs: string[];
  selectedInput?: Input;
}

export const midiContext = React.createContext<MIDIContextValue>(undefined);

export const MIDIProvider = ({ children }: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const enableMidi = async () => {
    try {
      await WebMidi.enable({ sysex: true });
      if (WebMidi.inputs.length) {
        dispatch(updateMIDIInput(WebMidi.inputs[0].name));
      }
      dispatch(updateMIDILoading(false));
    } catch (e) {
      console.error('Could not start Web MIDI', e);
      dispatch(updateMIDILoading(false));
      dispatch(updateMIDIError(true));
    }
  };

  useEffect(() => {
    enableMidi();
  }, []);

  if (state.loading) return;
  if (!state.loading && !state.error) {
    const inputs = WebMidi.inputs.length
      ? WebMidi.inputs.map((input) => input.name)
      : [];
    return (
      <midiContext.Provider
        value={{
          dispatch,
          inputs,
          selectedInput: state.input
            ? WebMidi.getInputByName(state.input)
            : undefined,
        }}
      >
        {children}
      </midiContext.Provider>
    );
  }
};
