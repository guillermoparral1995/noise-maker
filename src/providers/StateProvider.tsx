import React, { PropsWithChildren, useReducer } from 'react';
import initialState from '../store/initialState';
import reducer from '../store/reducer';
import { State } from '../types';
import { ActionTypes } from '../store/actions';

interface StateProviderValue {
  state: State;
  dispatch: React.Dispatch<ActionTypes>;
}

export const stateContext = React.createContext<StateProviderValue>(undefined);

export const StateProvider = ({ children }: PropsWithChildren) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  return (
    <stateContext.Provider value={{ state, dispatch }}>
      {children}
    </stateContext.Provider>
  );
};
