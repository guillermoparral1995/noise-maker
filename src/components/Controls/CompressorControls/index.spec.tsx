import {
  act,
  cleanup,
  fireEvent,
  render,
  screen,
} from '@testing-library/react';
import React from 'react';
import { CompressorControls_ as CompressorControls } from '.';
import {
  AudioContextMock,
  DynamicsCompressorNodeMock,
} from '../../../../__mocks__';
import { AudioContextProvider } from '../../../providers/AudioContextProvider';
import { Knobs } from '../../../types';
import { CompressorStateProvider } from './CompressorStateProvider';

const __mockGetInputByName = jest.fn();
jest.mock('webmidi', () => {
  return {
    ...jest.requireActual('webmidi'),
    WebMidi: {
      inputs: [],
      enable: jest.fn(),
      getInputByName: () => __mockGetInputByName(),
    },
  };
});

describe('Compressor controls', () => {
  afterEach(() => {
    cleanup();
  });

  const mockContext = new AudioContextMock();
  const mockCompressor = new DynamicsCompressorNodeMock({ ratio: 12 });

  it('should render correctly', async () => {
    const { findByTestId } = render(
      <AudioContextProvider
        __mocks={{ context: mockContext, compressor: mockCompressor }}
      >
        <CompressorStateProvider>
          <div data-testid="compressor-controls">
            <CompressorControls></CompressorControls>
          </div>
        </CompressorStateProvider>
      </AudioContextProvider>,
    );

    const controls = await findByTestId('compressor-controls');
    expect(controls).toMatchSnapshot();
  });

  it('should adjust param when knob is moved', async () => {
    render(
      <AudioContextProvider
        __mocks={{ context: mockContext, compressor: mockCompressor }}
      >
        <CompressorStateProvider>
          <CompressorControls></CompressorControls>
        </CompressorStateProvider>
      </AudioContextProvider>,
    );

    const knob = screen.getByTestId(Knobs.COMPRESSOR_RATIO);
    act(() => {
      knob.focus();
    });
    fireEvent.keyDown(knob, { key: 'ArrowUp', code: 'ArrowUp' });
    expect(mockCompressor.ratio.value).toEqual(12.01);
  });
});
