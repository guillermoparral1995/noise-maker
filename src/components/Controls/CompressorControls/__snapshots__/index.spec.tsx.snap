// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compressor controls should render correctly 1`] = `
<div
  data-testid="compressor-controls"
>
  <h3>
    Compressor
  </h3>
  <div
    class="controls_row"
  >
    <div
      data-testid="COMPRESSOR_THRESHOLD"
      tabindex="-1"
    >
      <p
        class="label"
      >
        Threshold
      </p>
      <webaudio-knob
        bodycolor="#f0f0f0"
        colors="mediumpurple;#f0f0f0;#ffffff;"
        defvalue="-24"
        diameter="35"
        highlightcolor="#ffffff"
        id="COMPRESSOR_THRESHOLD"
        indicatorcolor="mediumpurple"
        max="0"
        min="-100"
        step="0.01"
        style="height: 35px;"
        value="-24"
      />
      <p
        class="label"
      >
        -24
      </p>
    </div>
    <div
      data-testid="COMPRESSOR_RATIO"
      tabindex="-1"
    >
      <p
        class="label"
      >
        Ratio
      </p>
      <webaudio-knob
        bodycolor="#f0f0f0"
        colors="mediumpurple;#f0f0f0;#ffffff;"
        defvalue="12"
        diameter="35"
        highlightcolor="#ffffff"
        id="COMPRESSOR_RATIO"
        indicatorcolor="mediumpurple"
        max="20"
        min="1"
        step="0.01"
        style="height: 35px;"
        value="12"
      />
      <p
        class="label"
      >
        12
      </p>
    </div>
    <div
      data-testid="COMPRESSOR_KNEE"
      tabindex="-1"
    >
      <p
        class="label"
      >
        Knee
      </p>
      <webaudio-knob
        bodycolor="#f0f0f0"
        colors="mediumpurple;#f0f0f0;#ffffff;"
        defvalue="30"
        diameter="35"
        highlightcolor="#ffffff"
        id="COMPRESSOR_KNEE"
        indicatorcolor="mediumpurple"
        max="40"
        min="0"
        step="0.01"
        style="height: 35px;"
        value="30"
      />
      <p
        class="label"
      >
        30
      </p>
    </div>
    <div
      data-testid="COMPRESSOR_ATTACK"
      tabindex="-1"
    >
      <p
        class="label"
      >
        Attack
      </p>
      <webaudio-knob
        bodycolor="#f0f0f0"
        colors="mediumpurple;#f0f0f0;#ffffff;"
        defvalue="0.003"
        diameter="35"
        highlightcolor="#ffffff"
        id="COMPRESSOR_ATTACK"
        indicatorcolor="mediumpurple"
        max="1"
        min="0"
        step="0.01"
        style="height: 35px;"
        value="0.003"
      />
      <p
        class="label"
      >
        0.003
      </p>
    </div>
    <div
      data-testid="COMPRESSOR_RELEASE"
      tabindex="-1"
    >
      <p
        class="label"
      >
        Release
      </p>
      <webaudio-knob
        bodycolor="#f0f0f0"
        colors="mediumpurple;#f0f0f0;#ffffff;"
        defvalue="0.25"
        diameter="35"
        highlightcolor="#ffffff"
        id="COMPRESSOR_RELEASE"
        indicatorcolor="mediumpurple"
        max="1"
        min="0"
        step="0.01"
        style="height: 35px;"
        value="0.25"
      />
      <p
        class="label"
      >
        0.25
      </p>
    </div>
  </div>
</div>
`;
